### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by alecloudenback** [JuliaActuary/EconomicScenarioGenerators.jl#2] [General]
Implement common scenario generators:

Interest Rates
- Vasicek
- Cox-Ingersoll-Ross (CIR)
- Black-Derman-Toy
- Ho-Lee  
- Hull-White / Extended Hull White

Equity Models
- https://quant.stackexchange.com/questions/7663/stochastic-modeling-of-stock-price-process

Regulatory
- AAA ESG (https://github.com/JuliaActuary/Learn/blob/master/AAA_ESG.jl)
- Other? 

Output of these would be an intermediate representation for those who simply want the rates, but the recommended usage would be to then convert these into Yields.jl objects.

**Comment by kasperrisager** [comment]
Love to see this get going, I've wanted to grab the package name and start it for some time, but haven't gotten around to it.

A few loose thoughts about design
* The interes rate models here are arbitrage free, but let's make sure both P and Q measures are supported.
* I love the `collect` syntax, exactly how it should be done in Julia, imo. We could (maybe) go further and take the length out if the ESG and just use `take`.
* How would composibility work here? Say I want a model where my swap curve is Hull-White, but due to lack of data my inflation curve i Vasicek, I have a few equity indices and I want all the processes correlated?

**Reply by alecloudenback** [reply]
Do you mind expanding a bit on the first and 3rd point? I'm sort of learning as I go on this front and would like to understand a bit better about what you mean.

**Reply by kasperrisager** [reply]
1. Sorry about the P/Q lingo. What I mean is that the models you mention are arbitrage free interest rate models (which in other language means that they are martingales under the risk-neutral Q measure). That means they are good for valuation of financial products or technical provisions. It's just that some people who do a lot of valuation tend to think that economic scenario generators must always be arbitrage free, but that's not the case. There are lots of good use cases for scenarios that represent what you actually think will happen (= under the real-world P measure). I just wanted to make sure that absence-of-arbitrage does not get so deeply embedded that it's impossible to do the other things.
3.  There would be a lot of good uses for a package where you can set up a model of more than just an interest rate curve. As a more actuarial example, if you have a savings contract (that, say, expires at age 65) which is invested in something risky, but where the issuer has guaranteed a minimum annual interest rate of zero, the valuation would involve both interest rates and the return on the risky asset (and maybe those two processes would be correlated). As products get more complex, there are countless ways you might want economic scenarios to be constructed. How does one make a package that gives the user the flexibility to compose economic scenarios at will? I have never seen a good solution to that problem, I was hoping we could reach a good one here 😄

**Reply by alecloudenback** [reply]
Thanks :) 

Re #1 my simple understanding is that if the ESG can be used to replicate the price of a set of cashfows - so by constructing `HullWhite` using a Yields.jl curve (see example in README) then it's market-consistent. So it's more about how the ESG model is constructed than being able to convert a given model from RW to RN? Would you agree with that or do I misunderstand?

Re #2 I agree with the goal but don't have a clarity on how to achieve it. I think by adding the AAA/SOA's ESG which has correlations (I think?) then I can work to understand by example. And then figure out the right abstractions.

**Reply by kasperrisager** [reply]
#1, yes, it's mostly about not constraining the package too much - allowing `EconomicModel` to be pretty free (which it is currently) and not somehow structuring things so that only market consistent models will work.

#2, an existing model is probably a good place to start. Do you have a good reference for the AAA/SOA model? I'll also try to do some thinking myself instead of just throwing out ideas.
